include/rmm/mr/device/pool_memory_resource.hpp

```diff
@@ -181,18 +174,48 @@ class pool_memory_resource final
    */
   block_type expand_pool(size_t size, free_list& blocks, cudaStream_t stream)
   {
-    auto grow_size = size_to_grow(size);
-    if (grow_size == 0) {
-      RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE max pool size exceeded]",
-                    reinterpret_cast<void*>(stream),
-                    size);
-      RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
+    // if maximum size is not set
+    auto try_size =
+      maximum_pool_size_.has_value() ? size_to_grow(size) : std::max(size, pool_size());
+    while (try_size >= size) {
+      try {
+        auto b = block_from_upstream(try_size, stream);
+        current_pool_size_ += try_size;
+        return b;
+      } catch (std::exception& e) {
+        if (try_size == size) break;  // give up after we try size once
+        try_size = std::max(size, try_size / 2);
+      }
     }
-
-    current_pool_size_ += grow_size;
-    return block_from_upstream(grow_size, stream);
+    RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE max pool size exceeded]",
+                  reinterpret_cast<void*>(stream),
+                  size);
+    RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
```

Comment:
This code is a candidate for unit testing in isolation. If we template `expand_pool` to accept a `block_from_upstream` functor and extract it from the class, we'd be able to inject specific sizes in to the unit test. The class would call said extracted function.

```c++
template<typename GetBlockFunctor>
block_type expand_bool(GetBlockFunctor get_block, size_t size, freelist& blocks, cudaStream_t stream) {
  // ... real implementation as a pure function. can be tested in isolation ...
}

struct ... {
  block_type expand_bool(size_t size, freelist& blocks, cudaStream_t stream) {
    return expand_bool_static([this](...){ return this->block_from_upstream(...); }, size, blocks, stream);
  }
};
```

Reply:
I thought the wisdom was not to unit test implementation details, but rather interfaces.

Reply:
In any case, this function depends on the state of the class, so I don't think it can be extracted cleanly.

Reply:
> I thought the wisdom was not to unit test implementation details, but rather interfaces.

I thought you hadn't read any Uncle Bob. In that case, what interface are you testing to ensure this new combination of loop, arithmetic, state, bit-alignment, exception handling, and stream processing is correct?

Reply:
> I thought the wisdom was not to unit test implementation details, but rather interfaces.

This is true. But it appears we have a different understanding of this statement.

Tests should be written such that they contain no knowledge of the underlying implementation - only the expected inputs and outputs, and suspected edge cases (what should it do if I give it +/- inf, 0, or null?). 

This is testing an implementation detail:
```c++
int add_unless_special(int a, int b) { return a == 10 || b == 10 ? 0 : a + b; }

assert_eq(5, add_unless_special(3, 2));
assert_eq(0, add_unless_special(10, 10));
```
This is testing the same logic without regards to implementation details:
```c++
template<typename UnaryPredicate>
int add_unless_special_impl(UnaryPredicate is_special, int a, int b) {
  return is_special(a) || is_special(b) ? 0 : a + b;
  }
}

struct predicate
{
  int special_number;
  operator()(int x) { return x == special_number; }
}

struct constant_predicate
{
  bool result;
  operator()(int x) { return result; }
}

assert_eq(0, add_unless_special(predicate{3}, 3, 2));
assert_eq(0, add_unless_special(predicate{4}, 1, 4));
assert_eq(4, add_unless_special(predicate{7}, 8, -4));
assert_eq(0, add_unless_special(constant_predicate{false}, 1, 2));
assert_eq(3, add_unless_special(constant_predicate{true}, 1, 2));
assert_eq(0, add_unless_special(constant_predicate{false}, 3, 4));
assert_eq(7, add_unless_special(constant_predicate{true}, 3, 4));
```

The implementation detail is "a number is special if it is 10". The interface is "add these unless they are special". By injecting the meaning of "special" from within the test itself, we are no longer testing the meaning of "special", and instead only testing what happens if inputs are or are not special.

This is a contrived example, but when we take the same approach with something like `expand_pool`, we are able to test `expand_pool` as it's _own interface_, which is then used as an _implementation detail_ of `pool_memory_resource` (which can be tested without regard to `expand_pool`).
```
```

Reply:
I don't understand why a *private* function of a class that depends on instance data of that class is "too complex" if it can't be made static.  I think this request is distracting from the point of the PR.

Reply:
> > I thought the wisdom was not to unit test implementation details, but rather interfaces.
> 
> This is true. But it appears we have a different understanding of this statement.
> 
> Tests should be written such that they contain no knowledge of the underlying implementation - only the expected inputs and outputs, and suspected edge cases (what should it do if I give it +/- inf, 0, or null?).

The usual understanding of "black box" testing is that public interfaces are tested and ensure they behave in the way they are documented. Anything below that is usually considered an implementation detail. It would be impractical to test every internal function.

We already have tests that force pool growth and ensure the resource behaves as expected. That should be sufficient.

Reply:
I will try to add another test that sets no max and forces growth and if I can test this with an upstream that has a limited size I should be able to verify that growth succeeds and fails as expected.